src/filesystem.ts

```diff
@@ -0,0 +1,57 @@
+import localforage from 'localforage'
+
+import FileSystem from './fs'
+import * as dataRoot from './data-root'
+import { FS_CID, FS_TIMESTAMP , authenticatedUsername } from './common'
+
+
+/**
+ * Load a user's file system.
+ *
+ * @param username Optional, username of the user to load the file system from.
+ *                 Will try to load the file system of the authenticated user
+ *                 by default. Throws an error if there's no authenticated user.
+ */
+export async function loadFileSystem(username?: string): Promise<FileSystem> {
+  let cid, fs
+
+  // Determine username
+  username = username || (await authenticatedUsername() || undefined)
+  if (!username) throw new Error("User hasn't authenticated yet")
+
+  const lastLocalChange = await localforage.getItem(FS_TIMESTAMP) as number || 0
+  const currentTime = Date.now()
+
+  // If our last file-system change was over 15 minutes ago
+  if (currentTime - lastLocalChange > 30 * 60 * 1000) {
```

Comment:
This is done to avoid data loss, because either the DNS doesn't update fast enough, or the system/browser caches the DNS too long.

This could most likely be done better, so if anyone has any quick fixes for this, let me know.

Reply:
Yeah, time based caches cause all kinds of headaches. They’re generally considered a no-no in distributed systems (which the internet is).

In the long term, we’ll be able to check if you’re ahead of DNS because the CID will be in the history of your current file system.

In the meantime, why not track the root CIDs that are out of date in an array? i.e. When you load the page, record the FS root CID. Then when you update it, write that one down as well. Keep them in localStorage. When you try to pull from DNS, check that the hash isn’t in your local “out of date” cache. Now you’re not relying on time as a rough proxy, and can know instantly if there’s something new by using the Merkle proof (the method outlined here).

Reply:
Also, every time you successfully pull and update our local copy (i.e. DNS was ahead of your local), you can clear this cache. It doesn’t need to grow infinitely

Reply:
> when we implement FS versioning
👍 

> why not track the root CIDs that are out of date in an array?
Oh I like that, will use that.